storage/remote/queue_manager.go

```diff
@@ -302,11 +341,96 @@ func NewQueueManager(reg prometheus.Registerer, metrics *queueManagerMetrics, lo
 	}
 
 	t.watcher = wal.NewWatcher(reg, wal.NewWatcherMetrics(reg), logger, client.Name(), t, walDir)
+	if t.cfg.SendMetadata {
+		t.metadataWatcher = NewMetadataWatcher(logger, sm, client.Name(), t, t.cfg.MetadataSendInterval, flushDeadline)
+	}
 	t.shards = t.newShards()
 
 	return t
 }
 
+// AppendMetadata queues metadata to be sent to the remote storage. Metadata is sent all at once and is not parallelized.
+func (t *QueueManager) AppendMetadata(ctx context.Context, metadata []scrape.MetricMetadata) {
+	metadataSet := map[scrape.MetricMetadata]struct{}{}
+
+	// We create a set of the metadata by deduplicating based on the attributes of a
+	// scrape.MetricMetadata. In this case, a combination of metric name, help, type, and unit.
+	for _, entry := range metadata {
```

Comment:
Also maybe it's micro-optimization but we already loop for metadata once in collect so we could dedup there? This can have potentially ~100k elements.

Also we loop 3rd time, while we could just set the `prompb` into map straight away in `collect` as well.

Reply:
This actually makes sense now! Before, this lived in the `scrape` package so having `prompb` on that package implied introducing a concern not meant for that package. Now that this is part of the remote package it makes perfect sense to do it on the fly.